File: class1_ROS_Basic/act_sercli_example_pkg/src/simple_action_client_wait.py
#! /usr/bin/env python

import rospy
import actionlib
from custom_messages.msg import CounterAction, CounterGoal, CounterResult, CounterFeedback

# definition of the feedback callback. This will be called when feedback
# is received from the action server
# it just prints a message indicating a new message has been received
def feedback_callback(feedback):
    print('feedback :',feedback)

def result_callback(status,result):
    print('status : ',status)
    print('result :',result.count)

# initializes the action client node
rospy.init_node('counter_client')


action_server_name = '/counter_server'
client = actionlib.SimpleActionClient(action_server_name, CounterAction)
# waits until the action server is up and running
rospy.loginfo('Waiting for action Server '+action_server_name)
client.wait_for_server()
rospy.loginfo('Action Server Found...'+action_server_name)


# creates a goal to send to the action server
goal = CounterGoal()
goal.max_number = 10
goal.duration = 3
print(goal)


client.send_goal(goal, feedback_cb=feedback_callback, done_cb=result_callback)

rate = rospy.Rate(1)

rospy.loginfo("Lets Start The Wait for the Action To finish Loop...")
while not client.wait_for_result():
    rospy.loginfo("Doing Stuff while waiting for the Server to give a result....")
    
    rate.sleep()

rospy.loginfo("Example with WaitForResult Finished.")

File: class1_ROS_Basic/act_sercli_example_pkg_c/src/simple_action_client_wait.cpp
#include <custom_messages/CounterAction.h> // Note: "Action" is appended
#include <actionlib/client/simple_action_client.h>

void doneCb(const actionlib::SimpleClientGoalState& state,
            const custom_messages::CounterResultConstPtr& result)
{
  ROS_INFO("result : %d",result->count);
  ros::shutdown();
}

// Called once when the goal becomes active
void activeCb()
{
  ROS_INFO("Goal just went active");
}

void feedbackCb(const custom_messages::CounterFeedbackConstPtr& feedback)
{
  ROS_INFO("feedback : %f", feedback->percentage);
}

int main(int argc, char** argv)
{

  ros::init(argc, argv, "wait_for_result");
  
  actionlib::SimpleActionClient<custom_messages::CounterAction> client("counter_server", true);
  
  client.waitForServer();

  custom_messages::CounterGoal goal;
  goal.max_number = 10;
  goal.duration = 3;
  
  client.sendGoal(goal, &doneCb, &activeCb, &feedbackCb);
  //client.waitForResult();
  
  ros::Rate loop_rate(2);
    
  while (client.waitForResult() != true)
  {
    ROS_INFO("Doing Stuff while waiting for the Server to give a result...");
    loop_rate.sleep();
  }
  
  return 0;
}